flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/calcite/FlinkRelFactories.scala

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.calcite
+
+import org.apache.flink.table.plan.nodes.calcite.RankType.RankType
+import org.apache.flink.table.plan.nodes.calcite.{LogicalExpand, LogicalRank, LogicalSink, RankRange}
+import org.apache.flink.table.sinks.TableSink
+
+import org.apache.calcite.plan.Contexts
+import org.apache.calcite.rel.`type`.{RelDataType, RelDataTypeField}
+import org.apache.calcite.rel.{RelCollation, RelNode}
+import org.apache.calcite.rex.RexNode
+import org.apache.calcite.tools.RelBuilderFactory
+import org.apache.calcite.util.ImmutableBitSet
+
+import java.util
+
+/**
+  * Contains factory interface and default implementation for creating various rel nodes.
+  */
+object FlinkRelFactories {
```

Comment:
What's the difference between this and `FlinkLogicalRelFactories`? Do we have to go through the factory when we want to create logical nodes?

Reply:
This class defines the RelNode factory for extended node(e.g. Rank, Expand),  and `FlinkLogicalRelFactories` defines the all `FlinkLogicalRel` factories. 
There is no mandatory requirement to use factory to create logical nodes. However it's more easy to create a logical node if using `RelBuilderFactory`. And a rule class can be applied on calcite logical RelNode or Flink logical RelNode if `RelBuilderFactory` instance as its constructor argument.